src/Microsoft.ML.TimeSeries/SrCnnTransformBase.cs

```diff
@@ -0,0 +1,319 @@
+            {
+                _host.AssertValue(input);
+                var srcGetter = input.GetGetter<TInput>(input.Schema[_inputColumnIndex]);
+                Action<long> pinger = (long rowPosition) =>
+                {
+                    TInput src = default;
+                    srcGetter(ref src);
+                    state.UpdateState(ref src, rowPosition, _parent.WindowSize > 0);
+                };
+                return pinger;
+            }
+
+            public void CloneState()
+            {
+                if (Interlocked.Increment(ref _parent.StateRefCount) > 1)
+                {
+                    State = (SrCnnStateBase)_parent.StateRef.Clone();
+                }
+            }
+
+            public ITransformer GetTransformer()
+            {
+                return _parent;
+            }
+        }
+
+        internal abstract class SrCnnStateBase : SequentialTransformerBase<TInput, VBuffer<Double>, TState>.StateBase
+        {
+            protected SrCnnTransformBase<TInput, TState> Parent;
+
+            private protected SrCnnStateBase() { }
+
+            private protected override void CloneCore(TState state)
+            {
+                base.CloneCore(state);
+                Contracts.Assert(state is SrCnnStateBase);
+            }
+
+            private protected SrCnnStateBase(BinaryReader reader) : base(reader)
+            {
+            }
+
+            internal override void Save(BinaryWriter writer)
+            {
+                base.Save(writer);
+            }
+
+            private protected override void SetNaOutput(ref VBuffer<double> dst)
+            {
+                var outputLength = Parent.OutputLength;
+                var editor = VBufferEditor.Create(ref dst, outputLength);
+
+                for (int i = 0; i < outputLength; ++i)
+                    editor.Values[i] = 0;
+
+                dst = editor.Commit();
+            }
+
+            private protected sealed override void TransformCore(ref TInput input, FixedSizeQueue<TInput> windowedBuffer, long iteration, ref VBuffer<double> dst)
+            {
+                var outputLength = Parent.OutputLength;
+                Host.Assert(outputLength >= 2);
+
+                var result = VBufferEditor.Create(ref dst, outputLength);
+                for (int i = 0; i < outputLength; ++i)
+                    result.Values[i] = Double.NaN;
+
+                SpectralResidual(input, windowedBuffer, ref result);
+
+                dst = result.Commit();
+            }
+
+            private protected sealed override void InitializeStateCore(bool disk = false)
+            {
+                Parent = (SrCnnTransformBase<TInput, TState>)ParentTransform;
+                //TODO: assert for value threshold
+            }
+
+            private protected override void LearnStateFromDataCore(FixedSizeQueue<TInput> data)
+            {
+                //TODO:
```

Comment:
Do these `TODO`s need to be addressed? Or is no-oping here the correct thing?

Reply:
Remove the redundant TODO here.